xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.auth.SdsSecretConfig;
+import io.envoyproxy.envoy.api.v2.auth.Secret;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource.ApiType;
+import io.envoyproxy.envoy.api.v2.core.ConfigSource;
+import io.envoyproxy.envoy.api.v2.core.GrpcService;
+import io.envoyproxy.envoy.api.v2.core.GrpcService.GoogleGrpc;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.service.discovery.v2.SecretDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.discovery.v2.SecretDiscoveryServiceGrpc.SecretDiscoveryServiceStub;
+import io.grpc.Internal;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollDomainSocketChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.unix.DomainSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * SDS client used by {@link SslContextProvider}s to get SDS updates from the SDS server. This is
+ * most likely a temporary implementation until merged with the XdsClient.
+ */
+// TODO(sanjaypujare): once XdsClientImpl is ready, merge with it and add retry logic
+@Internal
+final class SdsClient {
+  private static final Logger logger = Logger.getLogger(SdsClient.class.getName());
+  private static final String SECRET_TYPE_URL = "type.googleapis.com/envoy.api.v2.auth.Secret";
+
+  private final HashMap<String, HashSet<SecretWatcher>> watcherMap = new HashMap<>();
```

Comment:
Don't we want to reduce our dependency on Guava? I can use it otherwise

Reply:
SetMultimap is available since 2.0, it is fair to use it (especially in xds). imo it's better than custom SetMultimap because there are potential garbage/memory usage issue.

Reply:
with one client per resource we don't need any kind of Map only a set of Watchers.